hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -557,9 +564,20 @@ SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long
                     partitionKeyComponents[i] = perMapCandidate.get(attribute).eval(null, evalContext);
                 }
 
-                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
-                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
-                );
+                // Mimic calculation performed for IMap put/get operations
+                // in AbstractSerializationService.calculatePartitionHash.
+                // IMap partitioning strategy is passed there.
+                //
+                // We cannot pass AttributePartitioningStrategy because we do not full key object, but
+                // IDENTITY_PARTITIONING_STRATEGY on partitionKeyComponents will give the same result
+                // as AttributePartitioningStrategy would on a full key.
+                // For other IMap strategies we use IMap strategy (only Default is supported)
+                // or null which will use global strategy.
+                Data keyData = serializationService.toData(
```

Comment:
NPE for `__key = ?` when the argument is `null`

Reply:
fixed in https://github.com/hazelcast/hazelcast/pull/25105/commits/8d2a3ec012b6ac66560e535440b6ee747484dfd6

Reply:
After merging master I get NPEs from scan partition pruning:

```
java.lang.NullPointerException: key cannot be null
	at com.hazelcast.internal.util.Preconditions.checkNotNull(Preconditions.java:60) ~[classes/:?]
	at com.hazelcast.internal.partition.PartitionServiceProxy.getPartition(PartitionServiceProxy.java:89) ~[classes/:?]
	at com.hazelcast.jet.sql.impl.connector.map.SpecificPartitionsImapReaderPms.init(SpecificPartitionsImapReaderPms.java:82) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.init.ExecutionPlanBuilder.lambda$createExecutionPlans$f148ad04$1(ExecutionPlanBuilder.java:150) ~[classes/:?]
```

Reply:
should be fixed now